### PR TITLE
Design/angle

### PR DESCRIPTION
**Pull request by HaicoDorenbos**
Complete redesigning of the Angle class which includes:
- Solving bug related to equality check of Angle instances.
- No new class instance creation for the += and -= operators
- Removing dead functions that are also code clones
- Better documentation of functions and the class
- Writing test cases for the Angle class
- Some small renamings

This PR can only be merged once Synchronize/utils angle #1126 on the Roboteam_ai branch is merged.

### CODE REVIEW

File: include/roboteam_utils/Angle.h [hunk 1]
```diff
@@ -12,218 +12,117 @@ namespace rtt {
...
 
         /**
-         * @brief Constrains the angle between 0 and 2 pi
-         *
-         * @return Angle A copy of `*this`
+         * Rescale the angle value to the range [-PI, PI).
+         * @return A reference to this rescaled Angle instance.
          */
-        Angle constrain() noexcept;
+        Angle rescale() noexcept;
```
Comment (rolfvdhulst) [inline include/roboteam_utils/Angle.h]: I personally like the old name more. Rescaling sounds like you´re doing a transformation on the angle (e.g. angles to degrees or multiplying it by some number), which you are not . I like the names 'normalize' or 'constrain' better as they are clearer (atleast, to me) and a bit more clear in what the function is actually doing.

File: test/AngleTest.cpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    /* Apply equivalence partitioning as testing method (https://en.wikipedia.org/wiki/Equivalence_partitioning) with the following partitions:
+     * angle < -PI, -PI <= angle < 0, angle = 0, 0 < angle < PI, angle >= PI
+     * and extend these partitions with boundary value analysis. We do not test angle = -FLT_MAX or angle = FLT_MAX, since we expect things to go wrong for these test cases, but
+     * we test other big positive and negative value instead. */
+
+    // Partition tests for: angle < -PI
+    double testAngle = -372349.9654;  // A random large negative number
+    double expectedRescale = -2.120912;
+    double actualRescale = Angle(testAngle).getAngle();
+    ASSERT_TRUE(fabs(actualRescale - expectedRescale) < EPSILON);
```
Comment (rolfvdhulst) [inline test/AngleTest.cpp:18]: Use ASSERT_NEAR(actualRescale,ExpectedRescale,EPSILON), it works better for some edge cases (floating point numbers are difficult and google tests helps you here)

Comment (rolfvdhulst) [reply test/AngleTest.cpp:18]: You don't need to change this though, just remember in the future. I'm convinced the code works fine.

File: include/roboteam_utils/Angle.h [hunk 3]
```diff
@@ -12,218 +12,117 @@ namespace rtt {
...
-         *
-         * Explicitly constraints.
-         *
-         * @param scalar Scalar to set `this->angle` to
+         * Set the angle value of this Angle instance, which will be directly rescaled.
+         * @param scalar A double value which represent the new angle value. It is allowed to use a value outside the range [-PI, PI). However you should be cautious with float
+         * values close to -MAX_FLOAT and close to MAX_FLOAT.
+         * @return A reference to this updated Angle instance.
          */
         Angle &operator=(const double &scalar) noexcept;
```
Comment (rolfvdhulst) [inline include/roboteam_utils/Angle.h]: This is quite inconsistent with the behavior of the other -= operator which returns an instance and not a reference

File: include/roboteam_utils/Angle.h [hunk 4]
```diff
@@ -12,218 +12,117 @@ namespace rtt {
...
-        /**
-         * @brief Sets the current angle equal to `this->angle - other.angle`
-         *
-         * @param other Other angle to subtract from `this`
-         * @return Angle A copy of `*this`
+         * Subtract the other Angle from this Angle, and directly rescale the angle value.
+         * @param other The other Angle.
+         * @return A reference to this updated Angle instance.
          */
         Angle operator-=(const Angle &other) noexcept;
```
Comment (rolfvdhulst) [inline include/roboteam_utils/Angle.h]: Inconsistent with other -= operator.

**Review by rolfvdhulst — COMMENTED** [review]
LGTM mostly, some minor comments.
You should only fix the -= operator inconistency IMO

File: include/roboteam_utils/Angle.h [hunk 5]
```diff
@@ -12,218 +12,117 @@ namespace rtt {
...
-        /**
-         * @brief Sets the current angle equal to`this->angle + other.angle`
-         *
-         * @param other Other angle to add to `this`
-         * @return Angle A copy of `*this`
+         * Add the other Angle to this Angle, and directly rescale the angle value.
+         * @param other The other Angle.
+         * @return A reference to this updated Angle instance.
          */
         Angle operator+=(const Angle &other) noexcept;
```
Comment (timovdk) [inline include/roboteam_utils/Angle.h]: Should actually return a reference

File: include/roboteam_utils/Angle.h (hunk 4 above)
Comment (timovdk) [reply include/roboteam_utils/Angle.h]: Yes, operators should return references

File: include/roboteam_utils/Angle.h (hunk 3 above)
Comment (timovdk) [reply include/roboteam_utils/Angle.h]: Also, why is there no `Angle &operator=(const Angle &angle);`?

File: include/roboteam_utils/Angle.h [hunk 6]
```diff
@@ -12,218 +12,117 @@ namespace rtt {
...
-         */
-        double angle;
+        double angle; // The current angle value, which is a value in the range [-PI, PI).
 
         /**
-         * @brief Constrains the angle between 0 and 2 pi
-         *
-         * @return Angle A copy of `*this`
+         * Rescale the angle value to the range [-PI, PI).
+         * @return A reference to this rescaled Angle instance.
```
Comment (timovdk) [inline include/roboteam_utils/Angle.h]: You are not returning a reference

File: include/roboteam_utils/Angle.h [hunk 7]
```diff
@@ -12,218 +12,117 @@ namespace rtt {
...
-
-        /**
-         * @brief Construct a new Angle object
-         *
-         * @param angle Angle to set the initial angle to
+         * Construct a new Angle instance by using a double value representing the angle value, which will be directly rescaled. However you should be cautious with float values
+         * close to -MAX_FLOAT and close to MAX_FLOAT.
+         * @param angle The given double value. It is allowed to use a value outside the range [-PI, PI).
          */
         Angle(double angle);
```
Comment (timovdk) [inline include/roboteam_utils/Angle.h]: Is there a reason this is not `Angle(const double &angle)`?

File: include/roboteam_utils/Angle.h (hunk 1 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/Angle.h]: Well normalize sound better to me than constrain

File: include/roboteam_utils/Angle.h (hunk 7 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/Angle.h]: No

File: include/roboteam_utils/Angle.h (hunk 3 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/Angle.h]: Well, since it is not used...

**Comment by HaicoDorenbos** [comment]
All of your mentioned feedback has been used :+1:

File: include/roboteam_utils/Angle.h [hunk 8]
```diff
@@ -12,218 +12,117 @@ namespace rtt {
...
         Angle() = default;
 
         /**
-         * @brief Copy constructor
+         * Construct a new Angle instance by using a double value representing the angle value, which will be directly normalized. However you should be cautious with float values
+         * close to -MAX_FLOAT and close to MAX_FLOAT.
+         * @param angle The given double value. It is allowed to use a value outside the range [-PI, PI).
          */
-        Angle(const Angle &copy) = default;
+        Angle(const double angle);
```
Comment (timovdk) [inline include/roboteam_utils/Angle.h:32]:
```suggestion
        Angle(const double &angle);
```

File: src/Angle.cpp [hunk 9]
```diff
@@ -1,15 +1,10 @@
...
 #include "../include/roboteam_utils/Angle.h"
 #include "../include/roboteam_utils/Vector2.h"
 #include "../include/roboteam_utils/Definitions.h"
-namespace rtt {
 
-    Angle::Angle(double angle)
-        : angle(angle) {
-        this->constrain();
+namespace rtt {
+    Angle::Angle(double angle) : angle(angle) {
```
Comment (timovdk) [inline src/Angle.cpp]:
```suggestion
    Angle::Angle(const double &angle) : angle(angle) {
```

Comment (HaicoDorenbos) [reply src/Angle.cpp]: Why a reference to a double value?

Comment (timovdk) [reply src/Angle.cpp]: to prevent a copy.
I think you also removed the copy constructor which is probably not what you intended to do

**Comment by HaicoDorenbos** [comment]
And do we all like the extensive equivalence partitioning testing technique with boundary value analysis?

File: include/roboteam_utils/Angle.h (hunk 8 above)
Comment (timovdk) [inline include/roboteam_utils/Angle.h]:
```suggestion
        Angle(const double &angle);
```

File: src/Angle.cpp (hunk 9 above)
Comment (HaicoDorenbos) [reply src/Angle.cpp]: The copy constructor was not used...

Comment (HaicoDorenbos) [reply src/Angle.cpp]: You want to copy it, since the angle gets normalized

**Review by timovdk — APPROVED** [review]

**Review by rttwente — APPROVED** [review]
